cirq/ops/fsim_gate.py

```diff
@@ -0,0 +1,137 @@
+
+    The unitary matrix of this gate is:
+
+        [[1, 0, 0, 0],
+         [0, a, b, 0],
+         [0, b, a, 0],
+         [0, 0, 0, c]]
+
+    where:
+
+        a = cos(theta)
+        b = i·sin(theta)
+        c = exp(i·phi)
+    """
+
+    def __init__(self, theta: float, phi: float):
+        self.theta = theta
+        self.phi = phi
+
+    def _value_equality_values_(self):
+        return self.theta, self.phi
+
+    def _is_parameterized_(self):
+        return cirq.is_parameterized(self.theta) or cirq.is_parameterized(
+            self.phi)
+
+    def _unitary_(self) -> Optional[np.ndarray]:
+        if cirq.is_parameterized(self):
+            return None
+        a = math.cos(self.theta)
+        b = 1j * math.sin(self.theta)
+        c = cmath.exp(1j * self.phi)
+        return np.array([
+            [1, 0, 0, 0],
+            [0, a, b, 0],
+            [0, b, a, 0],
+            [0, 0, 0, c],
+        ])
+
+    def _resolve_parameters_(self,
+                             param_resolver: 'cirq.ParamResolver'
+                             ) -> 'cirq.FSimGate':
+        return FSimGate(
+            cirq.resolve_parameters(self.theta, param_resolver),
+            cirq.resolve_parameters(self.phi, param_resolver))
+
+    def _apply_unitary_(self, args: 'cirq.ApplyUnitaryArgs'
+                       ) -> Optional[np.ndarray]:
+        if cirq.is_parameterized(self):
+            return None
+        if self.theta != 0:
+            inner_matrix = cirq.unitary(cirq.Rx(-2*self.theta))
+            oi = args.subspace_index(0b01)
+            io = args.subspace_index(0b10)
+            out = cirq.apply_matrix_to_slices(args.target_tensor,
+                                              inner_matrix,
+                                              slices=[oi, io],
+                                              out=args.available_buffer)
+        else:
+            out = args.target_tensor
+        if self.phi != 0:
+            ii = args.subspace_index(0b11)
+            out[ii] *= cmath.exp(1j * self.phi)
+        return out
+
+    def _decompose_(self, qubits) -> 'cirq.OP_TREE':
+        a, b = qubits
+        xx = cirq.XXPowGate(exponent=-self.theta / np.pi, global_shift=-0.5)
+        yy = cirq.YYPowGate(exponent=-self.theta / np.pi, global_shift=-0.5)
+        yield xx(a, b)
+        yield yy(a, b)
+        yield cirq.CZ(a, b)**(self.phi / np.pi)
+
+    def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs'):
+        t = _format_rads(args, self.theta)
+        p = _format_rads(args, self.phi)
+        return 'fsim({}, {})'.format(t, p), '#2'
+
+    def __pow__(self, power):
+        return FSimGate(
+            cirq.mul(self.theta, power),
+            cirq.mul(self.phi, power))
+
+    def __repr__(self):
+        return 'cirq.FSimGate(theta={}, phi={})'.format(
+            proper_repr(self.theta),
+            proper_repr(self.phi))
+
+
+def _format_rads(args: 'cirq.CircuitDiagramInfoArgs', radians: float) -> str:
```

Comment:
put this in a common place near CircuitDiagramINfoArgs?

Reply:
Acknowledged. Every time I write one of these it's slightly different. I'm waiting for a pattern to emerge.